Add subresource field to managed field entries

nodo opened the pull request:
#### What type of PR is this?

Add one of the following kinds:
/kind feature


#### What this PR does / why we need it:

While working on https://github.com/kubernetes/kubernetes/pull/98377 we decided that we needed a way to "namespace" managed fields of a subresource from the managed field of the main resource. This is especially useful when using the same manager to act on a subresource.

#### Special notes for your reviewer:

The generated code is in a separate commit.

#### Does this PR introduce a user-facing change?

```release-note
Managed fields of subresource objects have an additional field with the name of the subresource
```

nodo commented:
/cc @liggitt @apelisse

fedebongio commented:
/assign @apelisse 
/triage accepted

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -122,10 +122,10 @@ func decodeLiveManagedFields(liveObj runtime.Object) (Managed, error) {
 }
 
-func decodeManagedFields(liveObj, newObj runtime.Object, ignoreManagedFieldsFromRequestObject bool) (Managed, error) {
+func decodeManagedFields(liveObj, newObj runtime.Object, subresource string) (Managed, error) {
```
If this doesn't need to know the specific subresource used, I would keep the flag with the proper name as it gives more information about why this is used for.

nodo replied:
Good point, let me change it back.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/conflict.go:
```diff
@@ -75,11 +75,15 @@ func printManager(manager string) string {
 	}
+	managerStr := fmt.Sprintf("%q", encodedManager.Manager)
+	if encodedManager.Subresource != "" {
+		managerStr = fmt.Sprintf("%s in subresource %q", managerStr, encodedManager.Subresource)
```
nit: s/in/with/?

apelisse commented:
Looks good to me, we must think about the consequences of this change. @liggitt

nodo committed:
Add "subresource" field to ManagedFieldEntry

This field is useful to namespace the managed field entries of a
subresource and differentiate them from the ones of the main resource.

nodo committed:
Generated code

liggitt commented:
Chatted with @apelisse about this last week. Notes from that conversation:

Things we have to ensure:

1. 1.20 API servers will never choke on a managedFields list that drops the value from this field. Scenarios where that could occur:
    * Client reads from new API server, updates/patches to old API server
    * Client reads from new API server, server-side applies to old API server
2. 1.21 API servers will never choke on a managedFields list that drops the value from this field. Scenarios where that could occur:
    * Client writes subresource via 1.21 server, populating this field; reads from 1.20 server, dropping the field; writes to 1.21 server
    * Client reads from 1.21 server but drops the subresource field client-side, then writes back to 1.21 server

For handling of field ownership and conflicts, I see two categories:
1. fields that can be updated via multiple resource/subresource endpoints (e.g. labels in a resource that allows modifying metadata on status update, like pods)
    * the same manager applies the field via the main resource, then applies the same value via the subresource, then applies a change to the value via the main resource
    * the same manager applies the field via the main resource, then applies a different value via the subresource, then applies yet another different value via the main resource
2. fields that can only be updated via one resource or the other (like spec fields or status fields in a resource that divides access)


Consider and test (manually) what happens in the following scenario and make sure this doesn't regress a scenario that currently works:
* apply spec fields to spec, apply status fields to status (new server with this change), apply spec fields to spec (old server without this change)
* apply spec fields to spec, apply replica field via scale (new server with this change), apply spec fields to spec (old server without this change)

Consider and add tests for the following scenarios to make sure this doesn't regress scenarios that currently work:
* apply spec fields to spec, patch replica field via scale (new server with this change), apply spec fields to spec (old server without this change)
    ```
    kubectl apply --server-side -f manifest-without-replica.yaml
    kubectl scale -f manifest-without-replica.yaml --replicas=2
    kubectl apply --server-side -f new-manifest-without-replica.yaml
    ```
* apply a field that can be set via both regular and status subresource endpoints (like a label on a pod)

nodo commented:
Thanks a lot for the amazing feedback!

Just to keep you updated:

> 1.20 API servers will never choke on a managedFields list that drops the value from this field. 

[This test](https://github.com/kubernetes/kubernetes/pull/99210/commits/0ecf25b0f05a76cd2f61b1332f776c36b21e0021) should ensure that.

> 1.21 API servers will never choke on a managedFields list that drops the value from this field. 

[This test](https://github.com/kubernetes/kubernetes/pull/99210/commits/d1ba064f2f297fda3f9516f63a42d9574e108edd) should ensure that.

I will follow up with the manual tests you mentioned.

apelisse commented:
I think these three tests fairly well capture the problems that you mentioned in your comment @liggitt, could you PTAL?

apelisse commented:
/ok-to-test

nodo commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -2920,3 +2920,301 @@ spec:
+			}]
+		}
+	}`)
+	_, err := client.CoreV1().RESTClient().
```
Indeed, it's not needed, I have removed it.

nodo commented:
/retest

apelisse commented:
I couldn't find a test that shows that the subresource field is separating managers with the same name?

liggitt commented:
/assign
to add to my queue, though I'm happy to be preempted if another review has bandwidth

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1149,6 +1149,9 @@ type ManagedFieldsEntry struct {
 	// +optional
 	FieldsV1 *FieldsV1 `json:"fieldsV1,omitempty" protobuf:"bytes,7,opt,name=fieldsV1"`
+
+	// Subresource is the name of the subresource when the entry belongs to a subresource object.
```
an update here to explain how without this field an update to spec and an update to status causes pathological behavior would be a big help

nodo replied:
Thanks @deads2k. I was thinking to change the comment to:

```
Subresource is the name of the subresource when the entry belongs to a subresource object. This field allows the same manager to change a subresource or spec and to treat these operations as distinct workflows. In other words, a change to a subresource and a change to spec using the same managers will be considered as separate entries
```

WDYT?

apelisse replied:
What about that?
> Subresource is the name of the subresource used to update that object, or empty string if the object was updated through the main resource. The value of this field is used to distinguish between managers, even if they share the same name. For example, a status update will be distinct from a regular update using the same manager name.

nodo replied:
Much better, thanks @apelisse

nodo commented:
/retest

liggitt commented:
dual-assigning to surface in @lavalamp's review queue as well in case he frees up before me

apelisse commented:
Suite of tests looks good to me, "SubResource" comment has been improved to explain why it's needed and how it impacts workflows. lgtm for me (after rebase of course)

nodo commented:
/retest

deads2k commented:
Maybe I missed it, but I was expecting a direct test of applying to status and not losing objectmeta or a similar test.

the API change itself looks reasonable to me.

deads2k commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -2920,3 +2920,396 @@ spec:
+		t.Fatalf("Failed to apply: %v", err)
+	}
+
+	pod, err := client.CoreV1().Pods("default").Get(context.TODO(), "just-a-pod", metav1.GetOptions{})
```
forgot to check the labels themselves to be sure the values are right?

nodo replied:
Good point, let me add it.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1180,6 +1180,13 @@ type ManagedFieldsEntry struct {
+	// value of this field is used to distinguish between managers, even if they
+	// share the same name. For example, a status update will be distinct from a
+	// regular update using the same manager name.
+	Subresource string `json:"subresource,omitempty" protobuf:"bytes,8,opt,name=subresource"`
```
How would this work with a subresource that accepts two different versions, as is currently being added for pod/evict? (fortunately, that isn't the kind of subresource that matters for this)

lavalamp replied:
If subresource is non-empty, does it change the meaning of APIVersion? Does APIVersion remain about the parent resource's kind, or does it now refer to the subresource's kind?

apelisse replied:
Yeah, it's always about the parent resource. We do convert scale objects to parent resource version objects (and back) but the managedFields are always in parent resource version (otherwise we wouldn't be able to convert later)

lavalamp replied:
OK, that's right, because you've converted from whatever format the subresource was in already. OK. Comment could maybe mention this.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields_test.go:
```diff
@@ -462,6 +483,17 @@ func TestSortEncodedManagedFields(t *testing.T) {
+		{
+			name: "entries with subresource field",
+			managedFields: []metav1.ManagedFieldsEntry{
+				{Manager: "a", Operation: metav1.ManagedFieldsOperationApply, Subresource: "scale"},
```
maybe add in a second subresource to prove that those are sorted correctly?

nodo replied:
Makes sense 👍

lavalamp commented:
I agree that this needs a test that directly demonstrates the new subresource behavior. Other than that it looks pretty good to me.

nodo commented:
Thanks for the review @lavalamp @deads2k. I have address your comments and added an additional test to check the behaviour of the subresource field directly: [here](https://github.com/kubernetes/kubernetes/pull/99210/commits/56d0016abca6f690de705abdd1aabf611fb908e8#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dR3345-R3428). Let me know what you think. I could only test with "status", but I will add a similar test for "scale" once https://github.com/kubernetes/kubernetes/pull/98377 is merged.

nodo commented:
Added the "scale" test here: https://github.com/kubernetes/kubernetes/pull/98377/commits/156838cb424e85f7e4117d387858507fec1cc268#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13dL3408-R3444

lavalamp commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -2920,3 +2920,509 @@ spec:
+		t.Fatalf("Failed to update status: %v", err)
+	}
+
+	// TODO (nodo): add test for "scale" once we start tracking managed fields (#82046)
```
I am confused, I thought that's what this PR was addressing?

lavalamp replied:
OK yeah I think we should merge both together.
